### PR TITLE
Initial ExplorationWorker commit

### PR DESCRIPTION
**Pull request by tvmarino**
Initial ExplorationWorker commit. ExplorationWorker implements the
interactive clang compilation together with exploration for a
given module. This commit adds the class with its compile_module
function which will compile the module with a given compiler policy.

### CODE REVIEW

File: compiler_opt/rl/generate_bc_trajectories.py [hunk 1]
```diff
@@ -14,12 +14,101 @@
...
 from tf_agents.trajectories import policy_step
 from tf_agents.trajectories import time_step
+from tf_agents.specs import tensor_spec
+
+from compiler_opt.distributed import worker
+from compiler_opt.rl import corpus
+from compiler_opt.rl import env
+
+
+def add_int_feature(
```
Comment (mtrofin) [inline compiler_opt/rl/generate_bc_trajectories.py:34]: should this take a `type` parameter defaulted to np.int64? the type of the value should match the type of the element value in the feature list, right?

same q for the `add_float_feature`

File: compiler_opt/rl/generate_bc_trajectories.py [hunk 2]
```diff
@@ -14,12 +14,101 @@
...
+      the instance
+    feature_value: tf.int64 value of feature
+    feature_name: name of feature
+  """
+  f = sequence_example.feature_lists.feature_list[feature_name].feature.add()
+  lst = f.float_list.value
+  lst.extend([feature_value])
+
+
+def add_string_feature(
```
Comment (mtrofin) [inline compiler_opt/rl/generate_bc_trajectories.py:70]: why, do we ever think we'd use this?

File: compiler_opt/rl/generate_bc_trajectories.py [hunk 3]
```diff
@@ -102,3 +191,151 @@ def get_advice(self, state: time_step.TimeStep) -> np.ndarray:
...
+      mlgo_task: Type[env.MLGOTask],
+      use_greedy: bool = False,
+      exploration_frac: float = 1.0,
+      max_exploration_steps: int = 10,
+      exploration_policy_distr=None,
+      obs_action_specs: Optional[Tuple[time_step.TimeStep,
+                                       tensor_spec.BoundedTensorSpec,]] = None,
+      **kwargs,
+  ):
+    self.loaded_module_spec = loaded_module_spec
```
Comment (mtrofin) [inline compiler_opt/rl/generate_bc_trajectories.py:224]: isn't the coding convention that the member fields start with `_`?

File: compiler_opt/rl/generate_bc_trajectories.py [hunk 4]
```diff
@@ -102,3 +191,151 @@ def get_advice(self, state: time_step.TimeStep) -> np.ndarray:
...
+    self.use_greedy = use_greedy
+    if not obs_action_specs:
+      obs_spec = None
+      action_spec = None
+    else:
+      obs_spec = obs_action_specs[0].observation
+      action_spec = obs_action_specs[1]
+
+    if 'reward_key' not in kwargs:
```
Comment (mtrofin) [inline compiler_opt/rl/generate_bc_trajectories.py:233]: why not make `reward_key` an explicit parameter? if you type it appropriately (`str`, not `Optional[str]`), we can get a type checking error if the user doesn't pass it.

File: compiler_opt/rl/generate_bc_trajectories.py [hunk 5]
```diff
@@ -102,3 +191,151 @@ def get_advice(self, state: time_step.TimeStep) -> np.ndarray:
...
+      shutil.rmtree(working_dir_head)
+    if horizon <= 0:
+      raise ValueError(
+          f'Policy did not take any inlining decision for module {0}.'.format(
+              self.loaded_module_spec.name))
+    if curr_obs_dict.step_type != env.StepType.LAST:
+      raise ValueError(
+          f'Compilation loop exited at step type {0} before last step'.format(
+              curr_obs_dict.step_type))
+    native_size = curr_obs_dict.score_policy[self.reward_key]
```
Comment (mtrofin) [inline compiler_opt/rl/generate_bc_trajectories.py:298]: `native_size` is tied to inlining, perhaps `score`? same then below

File: compiler_opt/rl/generate_bc_trajectories.py [hunk 6]
```diff
@@ -102,3 +191,151 @@ def get_advice(self, state: time_step.TimeStep) -> np.ndarray:
...
+    native_size_list = np.float32(native_size) * np.float32(np.ones(horizon))
+    add_feature_list(sequence_example, native_size_list, 'reward')
```
Comment (mtrofin) [inline compiler_opt/rl/generate_bc_trajectories.py:300]: so I understand: the reward key could be `hi_mom`, but we rename it here to `reward`?

File: compiler_opt/rl/generate_bc_trajectories.py [hunk 7]
```diff
@@ -102,3 +191,151 @@ def get_advice(self, state: time_step.TimeStep) -> np.ndarray:
...
+    module_name_list = [self.loaded_module_spec.name for _ in range(horizon)]
+    add_feature_list(sequence_example, module_name_list, 'module_name')
```
Comment (mtrofin) [inline compiler_opt/rl/generate_bc_trajectories.py:302]: you may want to pull up in a global variable the feature names that are basically forming an API: IIUC `module_name` (1) can't be a feature in the original policy, and (2) is used as part of the implementation here.

You can also validate then that the given trace doesn't have this feature

File: compiler_opt/rl/generate_bc_trajectories.py [hunk 8]
```diff
@@ -102,3 +191,151 @@ def get_advice(self, state: time_step.TimeStep) -> np.ndarray:
...
+    self.exploration_frac = exploration_frac
+    self.max_exploration_steps = max_exploration_steps
+    self.exploration_policy_distr = exploration_policy_distr
+    logging.info('Reward key in exploration worker: %s', self.reward_key)
+
+  def compile_module(
+      self,
+      policy: Callable[[Optional[time_step.TimeStep]], np.ndarray],
+  ) -> tf.train.SequenceExample:
+    """Compiles the module with the given policy and outputs a seq. example.
```
Comment (mtrofin) [inline compiler_opt/rl/generate_bc_trajectories.py:260]: the documentation should capture that the sequence example also has some special, specifically-named features (`module_name`...) - IIUC

File: compiler_opt/rl/generate_bc_trajectories.py [hunk 9]
```diff
@@ -102,3 +191,151 @@ def get_advice(self, state: time_step.TimeStep) -> np.ndarray:
...
+      reward = np.array(curr_obs_dict.score_policy[self.reward_key])
+    else:
+      reward = np.array(0.)
+    curr_obs_step = step_type_converter[curr_obs_step]
+    timestep = time_step.TimeStep(
+        step_type=tf.convert_to_tensor([curr_obs_step],
+                                       dtype=tf.int32,
+                                       name='step_type'),
+        reward=tf.convert_to_tensor([reward], dtype=tf.float32, name='reward'),
+        discount=tf.convert_to_tensor([0.0], dtype=tf.float32, name='discount'),
```
Comment (mtrofin) [inline compiler_opt/rl/generate_bc_trajectories.py:323]: `reward` and `discount` here are features required by `TimeStep`, right? one thing that could clarify what feature names mean what is to maybe group them as static members into a couple of classes - just so there's a namespace-ing that helps readability (i.e. here it would be `name=TFAgentsTimeStep.DiscountFeatureName`)

Or, since there don't seem to be many names, just having globals and grouping them with comments also works.

Either way the reader should get sufficient hinting where these names come from or what the intent is with them.

**Review by mtrofin — COMMENTED** [review]
ptal at your commit message, probably formatting split it.

File: compiler_opt/rl/generate_bc_trajectories.py (hunk 1 above)
Comment (tvmarino) [reply compiler_opt/rl/generate_bc_trajectories.py:34]: The type value should match the type of the element value in the feature list. All of these are helper functions for add_feature_list. I can definitely make this take a type parameter but I don't see how this would help since at the end of the day I would have to cast whatever type is specified there to an int64 so that we create an int64 feature list. My understanding is that there are only a few types supported for feature lists and the three different add functions are these types.

File: compiler_opt/rl/generate_bc_trajectories.py (hunk 2 above)
Comment (tvmarino) [reply compiler_opt/rl/generate_bc_trajectories.py:70]: Imagine that during training we want to have specific weights only specific modules, then we can read off the module name from here and compute the relevant weight. It's something that we experimented with, however, it did not significantly improve on our final method. I can imagine this being useful in similar contexts though.

File: compiler_opt/rl/generate_bc_trajectories.py (hunk 6 above)
Comment (tvmarino) [reply compiler_opt/rl/generate_bc_trajectories.py:300]: The sequence_examples are what is going to be used for training. reward_key is the the key to be used in the environment to determine what we use as a reward, the two are somewhat independent.

File: compiler_opt/rl/generate_bc_trajectories.py (hunk 3 above)
Comment (tvmarino) [reply compiler_opt/rl/generate_bc_trajectories.py:224]: done

File: compiler_opt/rl/generate_bc_trajectories.py (hunk 4 above)
Comment (tvmarino) [reply compiler_opt/rl/generate_bc_trajectories.py:233]: done.

File: compiler_opt/rl/generate_bc_trajectories.py (hunk 1 above)
Comment (mtrofin) [reply compiler_opt/rl/generate_bc_trajectories.py:34]: ah, I see. sgtm

File: compiler_opt/rl/generate_bc_trajectories.py (hunk 2 above)
Comment (mtrofin) [reply compiler_opt/rl/generate_bc_trajectories.py:70]: if we don't currently (as in, part of this patch or known upcoming ones) use it, let's not add it. It's easy to add later.

File: compiler_opt/rl/generate_bc_trajectories.py (hunk 6 above)
Comment (mtrofin) [reply compiler_opt/rl/generate_bc_trajectories.py:300]: may be worth commenting that, easier for a reader to get that "it's intentional"

File: compiler_opt/rl/generate_bc_trajectories.py (hunk 9 above)
Comment (mtrofin) [reply compiler_opt/rl/generate_bc_trajectories.py:323]: just to make things easier for a reader. Rather than realizing that these may be tfagents things (not the most discoverable thing) and then reading on that, 1-2 sentences allows me to "get it" and move forward.

File: compiler_opt/rl/generate_bc_trajectories.py (hunk 2 above)
Comment (tvmarino) [reply compiler_opt/rl/generate_bc_trajectories.py:70]: I have used it in part of the training for some runs just not the ones that ended up giving the best policy. It would be more effort to remove this and the module_name feature than actually keep it at this point but really up to you.

File: compiler_opt/rl/generate_bc_trajectories.py (hunk 1 above)
Comment (mtrofin) [reply compiler_opt/rl/generate_bc_trajectories.py:34]: aaah! one thing I was missing: SequenceExamples really knows int64 and float32 IIRC, right? so the constraint here is what SequenceExamples knows about. 

May be worth stressing that somehow. `add_float_feature_to_seq_example` may be a mouthful, maybe just documenting that is sufficient (in the docstring) (i.e. that "int" is really int64 and always will be, because SeqExample. Then the reader gets it that it's not a point in time choice we made; the choice was made for us)

Comment (tvmarino) [reply compiler_opt/rl/generate_bc_trajectories.py:34]: Yeah exactly, I will add something to the documentation to emphasize this.

File: compiler_opt/rl/generate_bc_trajectories.py (hunk 8 above)
Comment (tvmarino) [reply compiler_opt/rl/generate_bc_trajectories.py:260]: done.

File: compiler_opt/rl/generate_bc_trajectories.py (hunk 5 above)
Comment (tvmarino) [reply compiler_opt/rl/generate_bc_trajectories.py:298]: changed it to reward

File: compiler_opt/rl/generate_bc_trajectories.py (hunk 6 above)
Comment (tvmarino) [reply compiler_opt/rl/generate_bc_trajectories.py:300]: See if the docstring for the function makes this more clear now.

File: compiler_opt/rl/generate_bc_trajectories.py (hunk 2 above)
Comment (mtrofin) [reply compiler_opt/rl/generate_bc_trajectories.py:70]: ok! just wanted to make sure it has an envisioned use (other than test)

File: compiler_opt/rl/generate_bc_trajectories.py (hunk 7 above)
Comment (tvmarino) [reply compiler_opt/rl/generate_bc_trajectories.py:302]: Check if this looks good, otherwise I haven't understood what you mean here.

Comment (tvmarino) [reply compiler_opt/rl/generate_bc_trajectories.py:302]: Check if the feature names already appear in what is returned by the env is done in _process_obs now.

File: compiler_opt/rl/generate_bc_trajectories.py [hunk 10]
```diff
@@ -111,6 +114,13 @@ def add_feature_list(seq_example: tf.train.SequenceExample,
     add_function(seq_example, feature, feature_name)
 
 
+@dataclasses.dataclass
+class SequenceExampleFeatureNames:
```
Comment (mtrofin) [inline compiler_opt/rl/generate_bc_trajectories.py:118]: nit: add a comment or doc string saying what this is for ("feature names that we depend on" for example).

also use these in the test

**Review by mtrofin — APPROVED** [review]

File: compiler_opt/rl/generate_bc_trajectories.py (hunk 10 above)
Comment (tvmarino) [reply compiler_opt/rl/generate_bc_trajectories.py:118]: done.
